Conflict with num-traits-0.2.15

I wanted to run the code from the website (such as [this](https://smartcorelib.org/user_guide/model_selection.html) ), but when I ran it, I found a conflict with num-trait-0.2.15. I'll attach the error below. 
`--> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/smartcore-0.2.0/src/linalg/evd.rs:474:35
     |
474  |                         z = p + z.copysign(p);
     |                                   ^^^^^^^^ multiple copysign found
     |
note: candidate #1 is defined in the trait num_traits::Float
    --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/num-traits-0.2.15/src/float.rs:1887:5
     |
1887 |     fn copysign(self, sign: Self) -> Self {
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in the trait RealNumber
    --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/smartcore-0.2.0/src/math/num.rs:27:5
     |
27   |     fn copysign(self, sign: Self) -> Self;
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: disambiguate the associated function for candidate #1
     |
474  |                         z = p + num_traits::Float::copysign(z, p);
     |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
help: disambiguate the associated function for candidate #2
     |
474  |                         z = p + RealNumber::copysign(z, p);
     |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0034]: multiple applicable items in scope
    --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/smartcore-0.2.0/src/linalg/evd.rs:573:64
     |
573  |                         let s = (p * p + q * q + r * r).sqrt().copysign(p);
     |                                                                ^^^^^^^^ multiple copysign found
     |
note: candidate #1 is defined in the trait num_traits::Float
    --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/num-traits-0.2.15/src/float.rs:1887:5
     |
1887 |     fn copysign(self, sign: Self) -> Self {
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in the trait RealNumber
    --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/smartcore-0.2.0/src/math/num.rs:27:5
     |
27   |     fn copysign(self, sign: Self) -> Self;
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: disambiguate the associated function for candidate #1
     |
573  |                         let s = num_traits::Float::copysign((p * p + q * q + r * r).sqrt(), p);
     |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
help: disambiguate the associated function for candidate #2
     |
573  |                         let s = RealNumber::copysign((p * p + q * q + r * r).sqrt(), p);
     |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0283]: type annotations needed
   --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/smartcore-0.2.0/src/linalg/evd.rs:524:29
    |
524 |                         y = T::from(0.75).unwrap() * s;
    |                             ^^^^^^^ cannot infer type for type {float}
    |
    = note: multiple impls satisfying {float}: ToPrimitive found in the num_traits crate:
            - impl ToPrimitive for f32;
            - impl ToPrimitive for f64;
note: required by a bound in num_traits::NumCast::from
   --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/num-traits-0.2.15/src/cast.rs:704:16
    |
704 |     fn from<T: ToPrimitive>(n: T) -> Option<Self>;
    |                ^^^^^^^^^^^ required by this bound in num_traits::NumCast::from
help: consider specifying the type argument in the function call
    |
524 |                         y = T::from::<T>(0.75).unwrap() * s;
    |                                    +++++

error[E0283]: type annotations needed
   --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/smartcore-0.2.0/src/linalg/evd.rs:525:29
    |
525 |                         x = T::from(0.75).unwrap() * s;
    |                             ^^^^^^^ cannot infer type for type {float}
    |
    = note: multiple impls satisfying {float}: ToPrimitive found in the num_traits crate:
            - impl ToPrimitive for f32;
            - impl ToPrimitive for f64;
note: required by a bound in num_traits::NumCast::from
   --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/num-traits-0.2.15/src/cast.rs:704:16
    |
704 |     fn from<T: ToPrimitive>(n: T) -> Option<Self>;
    |                ^^^^^^^^^^^ required by this bound in num_traits::NumCast::from
help: consider specifying the type argument in the function call
    |
525 |                         x = T::from::<T>(0.75).unwrap() * s;
    |                                    +++++

error[E0283]: type annotations needed
   --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/smartcore-0.2.0/src/linalg/evd.rs:526:29
    |
526 |                         w = T::from(-0.4375).unwrap() * s * s;
    |                             ^^^^^^^ cannot infer type for type {float}
    |
    = note: multiple impls satisfying {float}: ToPrimitive found in the num_traits crate:
            - impl ToPrimitive for f32;
            - impl ToPrimitive for f64;
note: required by a bound in num_traits::NumCast::from
   --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/num-traits-0.2.15/src/cast.rs:704:16
    |
704 |     fn from<T: ToPrimitive>(n: T) -> Option<Self>;
    |                ^^^^^^^^^^^ required by this bound in num_traits::NumCast::from
help: consider specifying the type argument in the function call
    |
526 |                         w = T::from::<T>(-0.4375).unwrap() * s * s;
    |                                    +++++

error[E0034]: multiple applicable items in scope
    --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/smartcore-0.2.0/src/linalg/svd.rs:119:35
     |
119  |                     g = -s.sqrt().copysign(f);
     |                                   ^^^^^^^^ multiple copysign found
     |
note: candidate #1 is defined in the trait num_traits::Float
    --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/num-traits-0.2.15/src/float.rs:1887:5
     |
1887 |     fn copysign(self, sign: Self) -> Self {
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in the trait RealNumber
    --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/smartcore-0.2.0/src/math/num.rs:27:5
     |
27   |     fn copysign(self, sign: Self) -> Self;
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: disambiguate the associated function for candidate #1
     |
119  |                     g = -num_traits::Float::copysign(s.sqrt(), f);
     |                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
help: disambiguate the associated function for candidate #2
     |
119  |                     g = -RealNumber::copysign(s.sqrt(), f);
     |                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0034]: multiple applicable items in scope
    --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/smartcore-0.2.0/src/linalg/svd.rs:155:35
     |
155  |                     g = -s.sqrt().copysign(f);
     |                                   ^^^^^^^^ multiple copysign found
     |
note: candidate #1 is defined in the trait num_traits::Float
    --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/num-traits-0.2.15/src/float.rs:1887:5
     |
1887 |     fn copysign(self, sign: Self) -> Self {
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in the trait RealNumber
    --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/smartcore-0.2.0/src/math/num.rs:27:5
     |
27   |     fn copysign(self, sign: Self) -> Self;
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: disambiguate the associated function for candidate #1
     |
155  |                     g = -num_traits::Float::copysign(s.sqrt(), f);
     |                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
help: disambiguate the associated function for candidate #2
     |
155  |                     g = -RealNumber::copysign(s.sqrt(), f);
     |                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0034]: multiple applicable items in scope
    --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/smartcore-0.2.0/src/linalg/svd.rs:302:59
     |
302  |                 f = ((x - z) * (x + z) + h * ((y / (f + g.copysign(f))) - h)) / x;
     |                                                           ^^^^^^^^ multiple copysign found
     |
note: candidate #1 is defined in the trait num_traits::Float
    --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/num-traits-0.2.15/src/float.rs:1887:5
     |
1887 |     fn copysign(self, sign: Self) -> Self {
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in the trait RealNumber
    --> /Users/user/.cargo/registry/src/github.com-1ecc6299db9ec823/smartcore-0.2.0/src/math/num.rs:27:5
     |
27   |     fn copysign(self, sign: Self) -> Self;
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: disambiguate the associated function for candidate #1
     |
302  |                 f = ((x - z) * (x + z) + h * ((y / (f + num_traits::Float::copysign(g, f))) - h)) / x;
     |                                                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
help: disambiguate the associated function for candidate #2
     |
302  |                 f = ((x - z) * (x + z) + h * ((y / (f + RealNumber::copysign(g, f))) - h)) / x;
     |                                                         ~~~~~~~~~~~~~~~~~~~~~~~~~~‰`

--- a/src/algorithm/neighbour/cover_tree.rs
+++ b/src/algorithm/neighbour/cover_tree.rs
@@ -65,7 +65,7 @@ struct Node<F: RealNumber> {
     max_dist: F,
     parent_dist: F,
     children: Vec<Node<F>>,
-    scale: i64,
+    _scale: i64,
 }
 
 #[derive(Debug)]
@@ -85,7 +85,7 @@ impl<T: Debug + PartialEq, F: RealNumber, D: Distance<T, F>> CoverTree<T, F, D> 
             max_dist: F::zero(),
             parent_dist: F::zero(),
             children: Vec::new(),
-            scale: 0,
+            _scale: 0,
         };
         let mut tree = CoverTree {
             base,
@@ -243,7 +243,7 @@ impl<T: Debug + PartialEq, F: RealNumber, D: Distance<T, F>> CoverTree<T, F, D> 
             max_dist: F::zero(),
             parent_dist: F::zero(),
             children: Vec::new(),
-            scale: 100,
+            _scale: 100,
         }
     }
 
@@ -304,7 +304,7 @@ impl<T: Debug + PartialEq, F: RealNumber, D: Distance<T, F>> CoverTree<T, F, D> 
                     max_dist: F::zero(),
                     parent_dist: F::zero(),
                     children,
-                    scale: 100,
+                    _scale: 100,
                 }
             } else {
                 let mut far: Vec<DistanceSet<F>> = Vec::new();
@@ -373,7 +373,7 @@ impl<T: Debug + PartialEq, F: RealNumber, D: Distance<T, F>> CoverTree<T, F, D> 
                         max_dist: self.max(consumed_set),
                         parent_dist: F::zero(),
                         children,
-                        scale: (top_scale - max_scale),
+                        _scale: (top_scale - max_scale),
                     }
                 }
             }

--- a/src/cluster/dbscan.rs
+++ b/src/cluster/dbscan.rs
@@ -155,11 +155,11 @@ impl<T: RealNumber + Sum, D: Distance<Vec<T>, T>> DBSCAN<T, D> {
         parameters: DBSCANParameters<T, D>,
     ) -> Result<DBSCAN<T, D>, Failed> {
         if parameters.min_samples < 1 {
-            return Err(Failed::fit(&"Invalid minPts".to_string()));
+            return Err(Failed::fit("Invalid minPts"));
         }
 
         if parameters.eps <= T::zero() {
-            return Err(Failed::fit(&"Invalid radius: ".to_string()));
+            return Err(Failed::fit("Invalid radius: "));
         }
 
         let mut k = 0;

--- a/src/cluster/kmeans.rs
+++ b/src/cluster/kmeans.rs
@@ -71,9 +71,9 @@ use crate::math::num::RealNumber;
 #[derive(Debug)]
 pub struct KMeans<T: RealNumber> {
     k: usize,
-    y: Vec<usize>,
+    _y: Vec<usize>,
     size: Vec<usize>,
-    distortion: T,
+    _distortion: T,
     centroids: Vec<Vec<T>>,
 }
 
@@ -208,9 +208,9 @@ impl<T: RealNumber + Sum> KMeans<T> {
 
         Ok(KMeans {
             k: parameters.k,
-            y,
+            _y: y,
             size,
-            distortion,
+            _distortion: distortion,
             centroids,
         })
     }
@@ -344,7 +344,7 @@ mod tests {
         let y = kmeans.predict(&x).unwrap();
 
         for i in 0..y.len() {
-            assert_eq!(y[i] as usize, kmeans.y[i]);
+            assert_eq!(y[i] as usize, kmeans._y[i]);
         }
     }
 

--- a/src/dataset/mod.rs
+++ b/src/dataset/mod.rs
@@ -67,14 +67,14 @@ pub(crate) fn serialize_data<X: RealNumber, Y: RealNumber>(
                 .data
                 .iter()
                 .copied()
-                .flat_map(|f| f.to_f32_bits().to_le_bytes().to_vec().into_iter())
+                .flat_map(|f| f.to_f32_bits().to_le_bytes().to_vec())
                 .collect();
             file.write_all(&x)?;
             let y: Vec<u8> = dataset
                 .target
                 .iter()
                 .copied()
-                .flat_map(|f| f.to_f32_bits().to_le_bytes().to_vec().into_iter())
+                .flat_map(|f| f.to_f32_bits().to_le_bytes().to_vec())
                 .collect();
             file.write_all(&y)?;
         }

--- a/src/ensemble/random_forest_classifier.rs
+++ b/src/ensemble/random_forest_classifier.rs
@@ -88,7 +88,7 @@ pub struct RandomForestClassifierParameters {
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[derive(Debug)]
 pub struct RandomForestClassifier<T: RealNumber> {
-    parameters: RandomForestClassifierParameters,
+    _parameters: RandomForestClassifierParameters,
     trees: Vec<DecisionTreeClassifier<T>>,
     classes: Vec<T>,
     samples: Option<Vec<Vec<bool>>>,
@@ -249,7 +249,7 @@ impl<T: RealNumber> RandomForestClassifier<T> {
         }
 
         Ok(RandomForestClassifier {
-            parameters,
+            _parameters: parameters,
             trees,
             classes,
             samples: maybe_all_samples,

--- a/src/ensemble/random_forest_regressor.rs
+++ b/src/ensemble/random_forest_regressor.rs
@@ -84,7 +84,7 @@ pub struct RandomForestRegressorParameters {
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[derive(Debug)]
 pub struct RandomForestRegressor<T: RealNumber> {
-    parameters: RandomForestRegressorParameters,
+    _parameters: RandomForestRegressorParameters,
     trees: Vec<DecisionTreeRegressor<T>>,
     samples: Option<Vec<Vec<bool>>>,
 }
@@ -215,7 +215,7 @@ impl<T: RealNumber> RandomForestRegressor<T> {
         }
 
         Ok(RandomForestRegressor {
-            parameters,
+            _parameters: parameters,
             trees,
             samples: maybe_all_samples,
         })

--- a/src/linalg/cholesky.rs
+++ b/src/linalg/cholesky.rs
@@ -87,8 +87,7 @@ impl<T: RealNumber, M: BaseMatrix<T>> Cholesky<T, M> {
         if bn != rn {
             return Err(Failed::because(
                 FailedError::SolutionFailed,
-                &"Can\'t solve Ax = b for x. Number of rows in b != number of rows in R."
-                    .to_string(),
+                "Can\'t solve Ax = b for x. Number of rows in b != number of rows in R.",
             ));
         }
 
@@ -128,7 +127,7 @@ pub trait CholeskyDecomposableMatrix<T: RealNumber>: BaseMatrix<T> {
         if m != n {
             return Err(Failed::because(
                 FailedError::DecompositionFailed,
-                &"Can\'t do Cholesky decomposition on a non-square matrix".to_string(),
+                "Can\'t do Cholesky decomposition on a non-square matrix",
             ));
         }
 
@@ -148,7 +147,7 @@ pub trait CholeskyDecomposableMatrix<T: RealNumber>: BaseMatrix<T> {
             if d < T::zero() {
                 return Err(Failed::because(
                     FailedError::DecompositionFailed,
-                    &"The matrix is not positive definite.".to_string(),
+                    "The matrix is not positive definite.",
                 ));
             }
 

--- a/src/linalg/evd.rs
+++ b/src/linalg/evd.rs
@@ -97,7 +97,7 @@ pub trait EVDDecomposableMatrix<T: RealNumber>: BaseMatrix<T> {
     }
 }
 
-fn tred2<T: RealNumber, M: BaseMatrix<T>>(V: &mut M, d: &mut Vec<T>, e: &mut Vec<T>) {
+fn tred2<T: RealNumber, M: BaseMatrix<T>>(V: &mut M, d: &mut [T], e: &mut [T]) {
     let (n, _) = V.shape();
     for (i, d_i) in d.iter_mut().enumerate().take(n) {
         *d_i = V.get(n - 1, i);
@@ -195,7 +195,7 @@ fn tred2<T: RealNumber, M: BaseMatrix<T>>(V: &mut M, d: &mut Vec<T>, e: &mut Vec
     e[0] = T::zero();
 }
 
-fn tql2<T: RealNumber, M: BaseMatrix<T>>(V: &mut M, d: &mut Vec<T>, e: &mut Vec<T>) {
+fn tql2<T: RealNumber, M: BaseMatrix<T>>(V: &mut M, d: &mut [T], e: &mut [T]) {
     let (n, _) = V.shape();
     for i in 1..n {
         e[i - 1] = e[i];
@@ -419,7 +419,7 @@ fn eltran<T: RealNumber, M: BaseMatrix<T>>(A: &M, V: &mut M, perm: &[usize]) {
     }
 }
 
-fn hqr2<T: RealNumber, M: BaseMatrix<T>>(A: &mut M, V: &mut M, d: &mut Vec<T>, e: &mut Vec<T>) {
+fn hqr2<T: RealNumber, M: BaseMatrix<T>>(A: &mut M, V: &mut M, d: &mut [T], e: &mut [T]) {
     let (n, _) = A.shape();
     let mut z = T::zero();
     let mut s = T::zero();
@@ -594,12 +594,7 @@ fn hqr2<T: RealNumber, M: BaseMatrix<T>>(A: &mut M, V: &mut M, d: &mut Vec<T>, e
                                 A.sub_element_mut(k + 1, j, p * y);
                                 A.sub_element_mut(k, j, p * x);
                             }
-                            let mmin;
-                            if nn < k + 3 {
-                                mmin = nn;
-                            } else {
-                                mmin = k + 3;
-                            }
+                            let mmin = if nn < k + 3 { nn } else { k + 3 };
                             for i in 0..mmin + 1 {
                                 p = x * A.get(i, k) + y * A.get(i, k + 1);
                                 if k + 1 != nn {
@@ -783,7 +778,7 @@ fn balbak<T: RealNumber, M: BaseMatrix<T>>(V: &mut M, scale: &[T]) {
     }
 }
 
-fn sort<T: RealNumber, M: BaseMatrix<T>>(d: &mut Vec<T>, e: &mut Vec<T>, V: &mut M) {
+fn sort<T: RealNumber, M: BaseMatrix<T>>(d: &mut [T], e: &mut [T], V: &mut M) {
     let n = d.len();
     let mut temp = vec![T::zero(); n];
     for j in 1..n {

--- a/src/linalg/evd.rs
+++ b/src/linalg/evd.rs
@@ -471,7 +471,7 @@ fn hqr2<T: RealNumber, M: BaseMatrix<T>>(A: &mut M, V: &mut M, d: &mut Vec<T>, e
                     A.set(nn, nn, x);
                     A.set(nn - 1, nn - 1, y + t);
                     if q >= T::zero() {
-                        z = p + z.copysign(p);
+                        z = p + RealNumber::copysign(z, p);
                         d[nn - 1] = x + z;
                         d[nn] = x + z;
                         if z != T::zero() {
@@ -570,7 +570,7 @@ fn hqr2<T: RealNumber, M: BaseMatrix<T>>(A: &mut M, V: &mut M, d: &mut Vec<T>, e
                                 r /= x;
                             }
                         }
-                        let s = (p * p + q * q + r * r).sqrt().copysign(p);
+                        let s = RealNumber::copysign((p * p + q * q + r * r).sqrt(), p);
                         if s != T::zero() {
                             if k == m {
                                 if l != m {

--- a/src/linalg/lu.rs
+++ b/src/linalg/lu.rs
@@ -46,13 +46,13 @@ use crate::math::num::RealNumber;
 pub struct LU<T: RealNumber, M: BaseMatrix<T>> {
     LU: M,
     pivot: Vec<usize>,
-    pivot_sign: i8,
+    _pivot_sign: i8,
     singular: bool,
     phantom: PhantomData<T>,
 }
 
 impl<T: RealNumber, M: BaseMatrix<T>> LU<T, M> {
-    pub(crate) fn new(LU: M, pivot: Vec<usize>, pivot_sign: i8) -> LU<T, M> {
+    pub(crate) fn new(LU: M, pivot: Vec<usize>, _pivot_sign: i8) -> LU<T, M> {
         let (_, n) = LU.shape();
 
         let mut singular = false;
@@ -66,7 +66,7 @@ impl<T: RealNumber, M: BaseMatrix<T>> LU<T, M> {
         LU {
             LU,
             pivot,
-            pivot_sign,
+            _pivot_sign,
             singular,
             phantom: PhantomData,
         }

--- a/src/linalg/mod.rs
+++ b/src/linalg/mod.rs
@@ -689,12 +689,11 @@ impl<'a, T: RealNumber, M: BaseMatrix<T>> Iterator for RowIter<'a, T, M> {
     type Item = Vec<T>;
 
     fn next(&mut self) -> Option<Vec<T>> {
-        let res;
-        if self.pos < self.max_pos {
-            res = Some(self.m.get_row_as_vec(self.pos))
+        let res = if self.pos < self.max_pos {
+            Some(self.m.get_row_as_vec(self.pos))
         } else {
-            res = None
-        }
+            None
+        };
         self.pos += 1;
         res
     }

--- a/src/linalg/naive/dense_matrix.rs
+++ b/src/linalg/naive/dense_matrix.rs
@@ -523,7 +523,6 @@ impl<T: RealNumber> PartialEq for DenseMatrix<T> {
         true
     }
 }
-
 impl<T: RealNumber> From<DenseMatrix<T>> for Vec<T> {
     fn from(dense_matrix: DenseMatrix<T>) -> Vec<T> {
         dense_matrix.values

--- a/src/linalg/svd.rs
+++ b/src/linalg/svd.rs
@@ -47,7 +47,7 @@ pub struct SVD<T: RealNumber, M: SVDDecomposableMatrix<T>> {
     pub V: M,
     /// Singular values of the original matrix
     pub s: Vec<T>,
-    full: bool,
+    _full: bool,
     m: usize,
     n: usize,
     tol: T,
@@ -116,7 +116,7 @@ pub trait SVDDecomposableMatrix<T: RealNumber>: BaseMatrix<T> {
                     }
 
                     let mut f = U.get(i, i);
-                    g = -s.sqrt().copysign(f);
+                    g = -RealNumber::copysign(s.sqrt(), f);
                     let h = f * g - s;
                     U.set(i, i, f - g);
                     for j in l - 1..n {
@@ -152,7 +152,7 @@ pub trait SVDDecomposableMatrix<T: RealNumber>: BaseMatrix<T> {
                     }
 
                     let f = U.get(i, l - 1);
-                    g = -s.sqrt().copysign(f);
+                    g = -RealNumber::copysign(s.sqrt(), f);
                     let h = f * g - s;
                     U.set(i, l - 1, f - g);
 
@@ -299,7 +299,7 @@ pub trait SVDDecomposableMatrix<T: RealNumber>: BaseMatrix<T> {
                 let mut h = rv1[k];
                 let mut f = ((y - z) * (y + z) + (g - h) * (g + h)) / (T::two() * h * y);
                 g = f.hypot(T::one());
-                f = ((x - z) * (x + z) + h * ((y / (f + g.copysign(f))) - h)) / x;
+                f = ((x - z) * (x + z) + h * ((y / (f + RealNumber::copysign(g, f))) - h)) / x;
                 let mut c = T::one();
                 let mut s = T::one();
 
@@ -428,13 +428,13 @@ impl<T: RealNumber, M: SVDDecomposableMatrix<T>> SVD<T, M> {
     pub(crate) fn new(U: M, V: M, s: Vec<T>) -> SVD<T, M> {
         let m = U.shape().0;
         let n = V.shape().0;
-        let full = s.len() == m.min(n);
+        let _full = s.len() == m.min(n);
         let tol = T::half() * (T::from(m + n).unwrap() + T::one()).sqrt() * s[0] * T::epsilon();
         SVD {
             U,
             V,
             s,
-            full,
+            _full,
             m,
             n,
             tol,

--- a/src/linear/linear_regression.rs
+++ b/src/linear/linear_regression.rs
@@ -94,7 +94,7 @@ pub struct LinearRegressionParameters {
 pub struct LinearRegression<T: RealNumber, M: Matrix<T>> {
     coefficients: M,
     intercept: T,
-    solver: LinearRegressionSolverName,
+    _solver: LinearRegressionSolverName,
 }
 
 impl LinearRegressionParameters {
@@ -155,7 +155,7 @@ impl<T: RealNumber, M: Matrix<T>> LinearRegression<T, M> {
 
         if x_nrows != y_nrows {
             return Err(Failed::fit(
-                &"Number of rows of X doesn\'t match number of rows of Y".to_string(),
+                "Number of rows of X doesn\'t match number of rows of Y",
             ));
         }
 
@@ -171,7 +171,7 @@ impl<T: RealNumber, M: Matrix<T>> LinearRegression<T, M> {
         Ok(LinearRegression {
             intercept: w.get(num_attributes, 0),
             coefficients: wights,
-            solver: parameters.solver,
+            _solver: parameters.solver,
         })
     }
 

--- a/src/linear/logistic_regression.rs
+++ b/src/linear/logistic_regression.rs
@@ -321,7 +321,7 @@ impl<T: RealNumber, M: Matrix<T>> LogisticRegression<T, M> {
 
         if x_nrows != y_nrows {
             return Err(Failed::fit(
-                &"Number of rows of X doesn\'t match number of rows of Y".to_string(),
+                "Number of rows of X doesn\'t match number of rows of Y",
             ));
         }
 

--- a/src/linear/ridge_regression.rs
+++ b/src/linear/ridge_regression.rs
@@ -96,7 +96,7 @@ pub struct RidgeRegressionParameters<T: RealNumber> {
 pub struct RidgeRegression<T: RealNumber, M: Matrix<T>> {
     coefficients: M,
     intercept: T,
-    solver: RidgeRegressionSolverName,
+    _solver: RidgeRegressionSolverName,
 }
 
 impl<T: RealNumber> RidgeRegressionParameters<T> {
@@ -226,7 +226,7 @@ impl<T: RealNumber, M: Matrix<T>> RidgeRegression<T, M> {
         Ok(RidgeRegression {
             intercept: b,
             coefficients: w,
-            solver: parameters.solver,
+            _solver: parameters.solver,
         })
     }
 

--- a/src/naive_bayes/categorical.rs
+++ b/src/naive_bayes/categorical.rs
@@ -161,7 +161,7 @@ impl<T: RealNumber> CategoricalNBDistribution<T> {
         let y_max = y
             .iter()
             .max()
-            .ok_or_else(|| Failed::fit(&"Failed to get the labels of y.".to_string()))?;
+            .ok_or_else(|| Failed::fit("Failed to get the labels of y."))?;
 
         let class_labels: Vec<T> = (0..*y_max + 1)
             .map(|label| T::from(label).unwrap())

--- a/src/optimization/first_order/lbfgs.rs
+++ b/src/optimization/first_order/lbfgs.rs
@@ -8,6 +8,7 @@ use crate::optimization::first_order::{FirstOrderOptimizer, OptimizerResult};
 use crate::optimization::line_search::LineSearchMethod;
 use crate::optimization::{DF, F};
 
+#[allow(clippy::upper_case_acronyms)]
 pub struct LBFGS<T: RealNumber> {
     pub max_iter: usize,
     pub g_rtol: T,

--- a/src/optimization/mod.rs
+++ b/src/optimization/mod.rs
@@ -4,6 +4,7 @@ pub mod line_search;
 pub type F<'a, T, X> = dyn for<'b> Fn(&'b X) -> T + 'a;
 pub type DF<'a, X> = dyn for<'b> Fn(&'b mut X, &'b X) + 'a;
 
+#[allow(clippy::upper_case_acronyms)]
 #[derive(Debug, PartialEq)]
 pub enum FunctionOrder {
     SECOND,

--- a/src/preprocessing/categorical.rs
+++ b/src/preprocessing/categorical.rs
@@ -78,8 +78,7 @@ fn find_new_idxs(num_params: usize, cat_sizes: &[usize], cat_idxs: &[usize]) -> 
         .zip(
             repeats
                 .zip(offset)
-                .map(|(r, o)| iter::repeat(o).take(r))
-                .flatten(),
+                .flat_map(|(r, o)| iter::repeat(o).take(r)),
         )
         .map(|(idx, ofst)| idx + ofst)
         .collect();

--- a/src/svm/svc.rs
+++ b/src/svm/svc.rs
@@ -222,7 +222,7 @@ impl<T: RealNumber, M: Matrix<T>, K: Kernel<T, M::RowVector>> SVC<T, M, K> {
 
         if n != y.len() {
             return Err(Failed::fit(
-                &"Number of rows of X doesn\'t match number of rows of Y".to_string(),
+                "Number of rows of X doesn\'t match number of rows of Y",
             ));
         }
 

--- a/src/svm/svr.rs
+++ b/src/svm/svr.rs
@@ -212,7 +212,7 @@ impl<T: RealNumber, M: Matrix<T>, K: Kernel<T, M::RowVector>> SVR<T, M, K> {
 
         if n != y.len() {
             return Err(Failed::fit(
-                &"Number of rows of X doesn\'t match number of rows of Y".to_string(),
+                "Number of rows of X doesn\'t match number of rows of Y",
             ));
         }
 

--- a/src/tree/decision_tree_classifier.rs
+++ b/src/tree/decision_tree_classifier.rs
@@ -118,7 +118,7 @@ pub enum SplitCriterion {
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[derive(Debug)]
 struct Node<T: RealNumber> {
-    index: usize,
+    _index: usize,
     output: usize,
     split_feature: usize,
     split_value: Option<T>,
@@ -204,7 +204,7 @@ impl Default for DecisionTreeClassifierParameters {
 impl<T: RealNumber> Node<T> {
     fn new(index: usize, output: usize) -> Self {
         Node {
-            index,
+            _index: index,
             output,
             split_feature: 0,
             split_value: Option::None,
@@ -514,7 +514,7 @@ impl<T: RealNumber> DecisionTreeClassifier<T> {
         visitor: &mut NodeVisitor<'_, T, M>,
         n: usize,
         count: &[usize],
-        false_count: &mut Vec<usize>,
+        false_count: &mut [usize],
         parent_impurity: T,
         j: usize,
     ) {

--- a/src/tree/decision_tree_regressor.rs
+++ b/src/tree/decision_tree_regressor.rs
@@ -97,7 +97,7 @@ pub struct DecisionTreeRegressor<T: RealNumber> {
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[derive(Debug)]
 struct Node<T: RealNumber> {
-    index: usize,
+    _index: usize,
     output: T,
     split_feature: usize,
     split_value: Option<T>,
@@ -137,7 +137,7 @@ impl Default for DecisionTreeRegressorParameters {
 impl<T: RealNumber> Node<T> {
     fn new(index: usize, output: T) -> Self {
         Node {
-            index,
+            _index: index,
             output,
             split_feature: 0,
             split_value: Option::None,

